substra/sdk/compute_plan.py

```diff
@@ -0,0 +1,157 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from copy import deepcopy
+import math
+
+from substra.sdk import schemas, graph, exceptions
+
+
+def _insert_into_graph(tuple_graph, tuple_id, in_model_ids):
+    if tuple_id in tuple_graph:
+        raise exceptions.InvalidRequest("Two tuples cannot have the same id.", 400)
+    in_model_ids = in_model_ids or list()
+    tuple_graph[tuple_id] = in_model_ids
+
+
+def get_dependency_graph(spec: schemas._BaseComputePlanSpec):
+    """Get the tuple dependency graph and, for each type of tuple, a mapping table id/tuple.
+    """
+    tuple_graph = dict()
+    traintuples = dict()
+    if spec.traintuples:
+        for traintuple in spec.traintuples:
+            _insert_into_graph(
+                tuple_graph=tuple_graph,
+                tuple_id=traintuple.traintuple_id,
+                in_model_ids=traintuple.in_models_ids,
+            )
+            traintuples[traintuple.traintuple_id] = traintuple
+
+    aggregatetuples = dict()
+    if spec.aggregatetuples:
+        for aggregatetuple in spec.aggregatetuples:
+            _insert_into_graph(
+                tuple_graph=tuple_graph,
+                tuple_id=aggregatetuple.aggregatetuple_id,
+                in_model_ids=aggregatetuple.in_models_ids,
+            )
+            aggregatetuples[aggregatetuple.aggregatetuple_id] = aggregatetuple
+
+    compositetuples = dict()
+    if spec.composite_traintuples:
+        for compositetuple in spec.composite_traintuples:
+            assert not compositetuple.out_trunk_model_permissions.public
+            _insert_into_graph(
+                tuple_graph=tuple_graph,
+                tuple_id=compositetuple.composite_traintuple_id,
+                in_model_ids=[
+                    model
+                    for model in [
+                        compositetuple.in_head_model_id,
+                        compositetuple.in_trunk_model_id,
+                    ]
+                    if model
+                ],
+            )
+            compositetuples[compositetuple.composite_traintuple_id] = compositetuple
+
+    return tuple_graph, traintuples, aggregatetuples, compositetuples
+
+
+def auto_batching(spec, is_creation: bool = True, batch_size: int = 20):
+    """Auto batching of the compute plan tuples
+    """
+
+    # Create the dependency graph and get the dict
+    # of tuples by id
+    (
+        tuple_graph,
+        traintuples,
+        aggregatetuples,
+        composite_traintuples,
```

Comment:
This is just a naming convention, but I feel that naming these variables traintuples_idx would make it clearer that they are dict which keys are ids. Just naming them `traintuples` looks like a list.